src/library/Connect/WorkloadInformationProvider.cs

```diff
@@ -480,21 +480,33 @@ private int GetServicePortValue(V1ServicePort servicePort, V1Container container
                 var isInWorkloadNamespace = StringComparer.OrdinalIgnoreCase.Equals(endpoint.Metadata.Namespace(), workloadNamespace);
                 foreach (var subset in endpoint.Subsets)
                 {
-                    // For headless service, we only add entries that specify hostname,
-                    // if hostname its not specify there is no dsn to reach the replicate.
-                    var addresses = subset.Addresses?.Where(a => !string.IsNullOrWhiteSpace(a?.Hostname));
-                    if (addresses == null || !addresses.Any())
+                    if (subset.Addresses == null)
                     {
                         continue;
                     }
-
-                    foreach (var address in addresses)
+                    
+                    foreach (var address in  subset.Addresses)
                     {
+                        if (address == null) {
+                            continue;
+                        }
+                        string dns = "";
+                        // If hostname is empty for the address, then dns used is that of the endpoint
+                        if (!string.IsNullOrWhiteSpace(address.Hostname))
+                        {
+                            dns = isInWorkloadNamespace ?
+                                    $"{address.Hostname}.{endpoint.Metadata.Name}" :
+                                    $"{address.Hostname}.{endpoint.Metadata.Name}.{endpoint.Metadata.Namespace()}";
+                        }
+                        else
+                        {
+                            dns = isInWorkloadNamespace ? 

```

Comment:
can you add UT for this if/else code block? ex: Assert on the List<EndpointInfo> using CompareLists or for each of this and compare the values ?

Reply:
done